driver/normalizer/normalizer.go

```diff
@@ -772,3 +770,180 @@ func (op dropNils) Check(st *State, n nodes.Node) (bool, error) {
 func (op dropNils) Construct(st *State, n nodes.Node) (nodes.Node, error) {
 	return op.op.Construct(st, n)
 }
+
+var (
+	typeGroup     = uast.TypeOf(uast.Group{})
+	typeFuncGroup = uast.TypeOf(uast.FunctionGroup{})
+)
+
+// triviaField specified a field with an array to put trivias into.
+var triviaField = map[string]string{
+	"Block":           "Statements",
+	"CompilationUnit": "Members",
+}
+
+// opMoveTrivias cuts trivia nodes from LeadingTrivia/TrailingTrivia fields
+// and wraps the node into uast:Group that contains those trivias.
+type opMoveTrivias struct {
+	sub Op
+}
+
+func (op opMoveTrivias) Kinds() nodes.Kind {
+	return nodes.KindObject
+}
+
+func (op opMoveTrivias) Check(st *State, n nodes.Node) (bool, error) {
+	obj, ok := n.(nodes.Object)
+	if !ok {
+		return false, nil
+	}
+	cloned := false
+	leading, ok1 := obj["LeadingTrivia"].(nodes.Array)
+	trailing, ok2 := obj["TrailingTrivia"].(nodes.Array)
+	if ok1 || ok2 {
+		// we saved the leading and trailing trivias, now wipe them from the node
+		obj = obj.CloneObject()
+		cloned = true
+		delete(obj, "LeadingTrivia")
+		delete(obj, "TrailingTrivia")
+	}
+
+	// move comments out of keywords and tokens into actual AST nodes
+	// TODO(dennwc): handle more cases:
+	//				 - Modifiers?
+	for key, sub := range obj {
+		if key != "ReturnType" &&
+			!strings.HasSuffix(key, "Token") &&
+			!strings.HasSuffix(key, "Keyword") {
+			continue
+		}
+		if uast.TypeOf(sub) != typeGroup {
+			continue
+		}
+		group := sub.(nodes.Object)
+		arr, ok := group["Nodes"].(nodes.Array)
+		if !ok {
+			continue
+		}
+		// we already joined trivias and the node into a single array,
+		// so we will have to find its index now
+		ind := -1
+		for i, sub := range arr {
+			if !strings.HasSuffix(uast.TypeOf(sub), "Trivia") {
+				ind = i
+				break
+			}
+		}
+		var node nodes.Node
+		if ind < 0 {
+			// cannot determine an index - pretend everything is a leading trivia
+			// TODO(dennwc): find another way if it happens in practice
+			leading = append(leading.CloneList(), arr...)
+		} else {
+			node = arr[ind]
+			leading = append(leading.CloneList(), arr[:ind]...)
+			trailing = append(arr[ind+1:len(arr):len(arr)], trailing...)
+		}
+		if !cloned {
+			obj = obj.CloneObject()
+			cloned = true
+		}
+		obj[key] = node
+	}
+
+	if len(leading) == 0 && len(trailing) == 0 {
+		if !cloned {
+			return false, nil // unmodified
+		}
+		// removed the trivia fields
+		return op.sub.Check(st, obj)
+	}
+
+	if field, ok := triviaField[uast.TypeOf(obj)]; ok {
+		// instead of wrapping the node, join trivias to a specified field
+		old, ok := obj[field].(nodes.Array)
+		if !ok {
+			return false, fmt.Errorf("expected %q field to be an array", field)
+		}
+		arr := make(nodes.Array, 0, len(leading)+len(old)+len(trailing))
+		arr = append(arr, leading...)
+		arr = append(arr, old...)
+		arr = append(arr, trailing...)
+
+		obj[field] = arr
```

Comment:
I suggest `return op.sub.Check(st, obj)` here so that the else branch can be outdented. There's a lot of nesting in this check.